File: python/leetcode/148.py
class Solution(object):
    def sortList(self, head):
        if not head or not head.next: return head
        fast, slow = head.next, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        nhead = slow.next
        slow.next = None
        l1 = self.sortList(head)
        l2 = self.sortList(nhead)
        return self.merge(l1, l2)
    
    def merge(self, l1, l2):
        if not l1 or not l2: return l1 if l1 else l2
        head = ListNode(0)
        ptr = head
        while l1 and l2:
            if l1.val < l2.val:
                ptr.next = l1
                l1 = l1.next
            else:
                ptr.next = l2
                l2 = l2.next
            ptr = ptr.next
        ptr.next = l1 if l1 else l2
        return head.next

File: cpp/leetcode/148.cpp
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode *fast = head->next, *slow = head;
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }
        ListNode* nhead = slow->next;
        slow->next = NULL;
        fast = sortList(nhead);
        slow = sortList(head);
        ListNode *dummy = new ListNode(0);
        head = dummy;
        while(slow && fast){
            if(fast->val < slow->val) {head->next = fast; fast = fast->next;}
            else {head->next = slow; slow = slow->next;}
            head = head->next;
        }
        if(fast || slow) head->next = fast ? fast : slow;
        return dummy->next;
    }
};